Library/Homebrew/livecheck/strategy.rb

```diff
@@ -119,25 +120,58 @@ def self.page_headers(url)
 
       # Fetches the content at the URL and returns a hash containing the
       # content and, if there are any redirections, the final URL.
+      # If `curl` encounters an error, the hash will contain a `:messages`
+      # array with the error message instead.
       #
       # @param url [String] the URL of the content to check
       # @return [Hash]
       sig { params(url: String).returns(T::Hash[Symbol, T.untyped]) }
       def self.page_content(url)
         original_url = url
 
-        # Manually handling `URI#open` redirections allows us to detect the
-        # resolved URL while also supporting HTTPS to HTTP redirections (which
-        # are normally forbidden by `OpenURI`).
-        begin
-          content = URI.parse(url).open(redirect: false, &:read)
-        rescue OpenURI::HTTPRedirect => e
-          url = e.uri.to_s
-          retry
+        args = curl_args(
+          "--compressed",            # Request compressed transfer encoding
+          "--include",               # Include HTTP response headers in output
+          "--location",              # Follow redirects
+          "--connect-timeout", "10", # Max time allowed for connection (secs)
+          "--max-time", "15"         # Max time allowed for transfer (secs)
+        )
+
+        stdout, stderr, status = curl_with_workarounds(
+          *args, url,
+          print_stdout: false, print_stderr: false,
+          debug: false, verbose: false,
+          user_agent: :default, timeout: 20,
+          retry: false
+        )
+
+        unless status.success?
+          /^(?<error_msg>curl: \(\d+\) .+)/ =~ stderr
+          return { messages: [error_msg] } if error_msg.present?
         end
 
-        data = { content: content }
-        data[:final_url] = url unless url == original_url
+        # stdout contains the header information followed by the page content.
+        # We use #scrub here to avoid "invalid byte sequence in UTF-8" errors.
+        output = stdout.scrub
+
+        # Separate the headers/body and identify the final URL (after any
+        # redirections)
+        status_code = :unknown
+        while status_code == :unknown || status_code.to_s.start_with?("3")
```

Comment:
```suggestion
        status_code = :unknown
        max_redirects = 5
        redirects = 0
        while status_code == :unknown || status_code.to_s.start_with?("3")
          redirects += 1
          break if redirects > max_redirects

```

Reply:
This code is simply parsing curl's output and any HTTP requests have already been made at this point. Restricting the number of redirects allowed in the output doesn't restrict the number of redirections that curl actually follows. More importantly, this could stop the parsing of response heads before reaching the final response and unexpectedly leave head responses in the output. It's necessary to work through all the head responses to effectively isolate the body content.

If the intention is to restrict the number of redirections that curl follows, then we simply need to add `"--max-redirs", "5",` to the `curl_args` call.

Reply:
The intention is to avoid having a `while` loop that has a maximum number of iterations. If it makes sense to align this with `--max-redirs`: cool, let's do that.

Reply:
I made the conditions for the `while` loop more explicit (closer to what I'm doing in #11252) and added logic to `raise` if the loop does more than five iterations.

I've used `iterations` in the variable names instead of `redirects`, since `redirects` would equal `1` on the first iteration even if there is no redirection. Similarly, the last iteration when there are redirections wouldn't have an accurate number of redirections either (e.g., `redirects = 2` when there's a redirect response followed by a normal response). `redirects` wouldn't be an accurate indicator of the number of redirections encountered, so it's arguably better to use something generic like `iterations`.

Additionally, I've set this to raise a `Too many redirects (max = #{max_iterations})` message when we reach the maximum. This leads to the error being surfaced in the livecheck output (normal or JSON) and prevents livecheck from working with curl output (i.e., `data[:content]`) where we haven't properly isolated the body from the heads.